bindings/Python/py11Engine.cpp

```diff
@@ -129,6 +129,12 @@ void Engine::Get(Variable variable, pybind11::array &array, const Mode launch)
 #define declare_type(T)                                                        \
     else if (type == helper::GetDataType<T>())                                 \
     {                                                                          \
+        if (!array.dtype().is(pybind11::dtype::of<T>()))                       \
```

Comment:
If the type and size of received data is known, we could just reallocate the buffer to the received type.

Reply:
Yes, we have other options.  We could also redefine Get() so that it simply returned a buffer of the right type instead of asking the user to provide a buffer that might potentially be of the wrong type.  (See notes in issue #3538.)  The current structure is likely more the result of directly translating the C++ bindings into Python without a lot of thought of what a redesign should look like.  A redesign might be desirable, but not for this release.  For this, I did the minimum that would have let us know about a problem that came up in that issue.  Want to create an issue for rethinking the Python APIs?

Reply:
Probably it should be improved rather on the core level, not in the python binding.

Reply:
I'm happy to hear ideas about that, but I do think that Python has specific capabilities (and flexibilities) that aren't present in C++, C, and Fortran, specifically regarding type flexibility.  There will be things that we could do in Python that simply aren't easily replicated in type-safe languages (at least not efficiently) and those are the sorts of things that could be more easily addressed in bindings, and would likely be difficult to mirror all the way into core.